clippy_lints/src/functions.rs

```diff
@@ -268,6 +392,160 @@ impl<'a, 'tcx> Functions {
     }
 }
 
+fn check_needless_must_use(
+    cx: &LateContext<'_, '_>,
+    decl: &hir::FnDecl,
+    item_id: hir::HirId,
+    item_span: Span,
+    attr: &Attribute,
+) {
+    if returns_unit(decl) {
+        span_lint_and_then(
+            cx,
+            MUST_USE_UNIT,
+            function_header_span(cx, item_span),
+            "this unit-returning function has a `#[must_use]` attribute",
+            |db| {
+                db.span_suggestion(
+                    attr.span,
+                    "remove the attribute",
+                    "".into(),
+                    Applicability::MachineApplicable,
+                );
+            },
+        );
+    } else if !attr.is_value_str() && !in_external_macro(cx.sess(), item_span) && is_must_use_ty(cx, return_ty(cx, item_id)) {
+        span_help_and_lint(
+            cx,
+            DOUBLE_MUST_USE,
+            function_header_span(cx, item_span),
+            "this function has an empty `#[must_use]` attribute, but returns a type already marked as `#[must_use]`",
+            "either add some descriptive text or remove the attribute",
+        );
+    }
+}
+
+fn check_must_use_candidate<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    decl: &'tcx hir::FnDecl,
+    body: &'tcx hir::Body,
+    item_span: Span,
+    item_id: hir::HirId,
+    msg: &str,
+) {
+    if has_mutable_arg(cx, body)
+        || mutates_static(cx, body)
+        || in_external_macro(cx.sess(), item_span)
+        || returns_unit(decl)
+        || is_must_use_ty(cx, return_ty(cx, item_id))
+    {
+        return;
+    }
+    let fn_span = function_header_span(cx, item_span);
+    span_lint_and_then(cx, MUST_USE_CANDIDATE, fn_span, msg, |db| {
+        if let Some(snippet) = snippet_opt(cx, fn_span) {
+            db.span_suggestion(
+                fn_span,
+                "add the attribute",
+                format!("#[must_use] {}", snippet),
+                Applicability::MachineApplicable,
+            );
+        }
+    });
+}
+
+fn must_use_attr(attrs: &[Attribute]) -> Option<&Attribute> {
```

Comment:
I believe you can replace this function with:

```rust
syntax::attr::contains_name(attrs, sym!(must_use)
```

Reply:
That returns only `bool`, I need the attribute for its node and span later.